### PR TITLE
feat(oauth): thread ClientRef through registration and implement static clients

### PR DESCRIPTION
**Pull request by Prabhjot-Sethi**
## Summary

Implements **AUTH-0010** (Commit 2 of the OAuth Multi-Client Support proposal). Threads the consumer-defined opaque `ClientRef` through client registration and lookup, and provides the full implementation of `RegisterStaticClient` (previously a stub returning `errStaticRegistrationUnimplemented`). After this change, two clients for the same authorization server with different `ClientRef` values coexist independently.

## Changes

- **`findClient`** now accepts and uses `clientRef`, resolving clients by `(ServerURL, ClientRef)`.
- **`GetClient(ctx, serverURL, clientRef)`** — new signature; builds the composite `ClientKey`.
- **`RegisterDynamicClient` / `ReRegisterClient`** read `opts.ClientRef` and thread it through lookup, the per-server registration lock key, delete, and insert. Dynamic registration uses `ClientRef ""`; with the `omitempty` BSON tag the lock identity is unchanged, so existing dynamic serialization is preserved.
- **`RegisterStaticClient(ctx, serverURL, clientRef, entry)`** — full implementation:
  - rejects empty `clientRef` (`""` is reserved exclusively for the dynamic slot),
  - normalizes `serverURL`,
  - stamps `RegistrationType = "static"` and `RegisteredAt`,
  - upserts via `Table.Locate`, so `ClientSecret` / `RegistrationAccessToken` are encrypted at rest by `ClientEntry.MarshalBSON` (same path as dynamic).
  - Removed the now-unused `errStaticRegistrationUnimplemented` sentinel.
- **`token.go` / `authorization.go`** — updated the other `findClient` call sites for the new signature (`key.ClientRef` / `merged.ClientRef`).

## Tests

- Updated `TestGetClient_HitAndMiss` and the fake client store to model the composite `(ServerURL, ClientRef)` key (+ a `Locate` upsert).
- Replaced `TestRegisterStaticClient_Unimplemented` with positive coverage:
  1. **Two static clients, same server** (`tenant-a` / `tenant-b`) store and resolve independently.
  2. **Static + dynamic coexistence** — a dynamic client (`""`) and a static client (`tenant-a`) for one server do not interfere.
  3. **Empty `clientRef` rejected.**
  - Plus metadata/upsert and ref-distinguishing `GetClient` tests.

## Verification

- `go build ./...` ✅
- `go test -race ./oauth/...` ✅ (full suite `go test ./...` passes)
- `go vet ./oauth/`, `gofmt -l oauth/`, `golangci-lint run ./oauth/` — clean (0 issues)

Refs: AUTH-0010

## Release Notes

* **New Features**
  * OAuth system now supports provisioning and managing multiple distinct clients per server.
  * Static OAuth client registration is now fully implemented, enabling direct provisioning of client credentials.

* **Improvements**
  * Authorization, token refresh, and token revocation flows now correctly identify and use the appropriate client for each request based on client configuration.

### CODE REVIEW

**Review by dev-arya23 — APPROVED** [review]
## What I verified

- **Cross-repo dependency (`table.Table.Locate`)**: Verified at source in `go-core-stack/core/table/generic.go` — `Locate` exists and performs upsert via `UpdateOne(ctx, key, entry, true)`. The `clientStore` interface addition is correctly satisfied by the concrete table type.
- **Type consistency with PR #33 (AUTH-0009)**: Confirmed `ClientRef` is already present in `ClientKey`, `TokenKey`, `RegistrationLockKey`, `RegisterClientOptions`, `AuthorizeOptions`, and `PendingAuthState` on the `main` branch. This PR correctly threads it through the registration and lookup logic.
- **Backward compatibility**: Dynamic registration continues to use `ClientRef ""`. With `omitempty` on the BSON tag, existing MongoDB documents without `clientRef` will deserialize as `""`, so the composite key is backward-compatible with pre-existing dynamic clients.
- **Lock key scoping**: Lock key now includes `ClientRef`, so static and dynamic registrations for the same server do not serialize on each other — correct design.
- **Static registration correctness**:
  - `clientRef` rejection via `strings.TrimSpace` catches both empty and whitespace-only values — good defensive guard.
  - `RegistrationType = "static"` and `RegisteredAt` correctly stamped.
  - Upsert via `Locate` allows secret rotation without a separate delete — clean.
  - No distributed lock for static registration is correct: `Locate` is an idempotent upsert with no external network call, so last-write-wins is safe. The dynamic path's lock exists to serialize against the external RFC 7591 endpoint, which doesn't apply here.
- **All `findClient` call sites updated**: `registration.go` (5 sites), `token.go` (2 sites: `refreshTokenExchange`, `revokeAtServer`), `authorization.go` (1 site: `authorizationURL`) — all consistently pass `clientRef`.
- **`performRegistration`** correctly passes `clientRef` through to the `Insert` key.
- **Error path analysis**: All errors are correctly classified — `InvalidArgument` for validation failures (permanent), `Unknown`/`GetErrCode` wrapping for infrastructure failures, `NotFound` tolerance on delete. No transient/permanent confusion.
- **Test coverage**: Comprehensive — empty ref rejection, metadata stamping + upsert, two-tenant coexistence, static + dynamic coexistence, `GetClient` disambiguation by `clientRef`. The `ckey()` helper in the fake store correctly models the composite `(ServerURL, ClientRef)` key with `\x00` separator.

## Suggestions

**[SUGGESTION]**: Consider adding a `TestRegisterStaticClient_EmptyServerURLRejected` test to exercise the `normalizeServerURL` → empty validation path in `registerStaticClient`. The code handles it correctly (`normalizeServerURL` returns `""` → `InvalidArgument` error), but there's no explicit test for it alongside the `clientRef` rejection test. Minor — can be a follow-up.

Clean implementation. The API surface changes (`GetClient` and `RegisterStaticClient` signatures) are appropriate at this pre-release stage. This lands Commit 2 of the multi-client keying proposal cleanly.

File: oauth/authorization.go [hunk 1]
```diff
@@ -108,7 +108,7 @@ func authorizationURL(ctx context.Context, servers serverCache, clients clientSt
 
 	// The client must already be registered and the server already discovered:
 	// the consumer drives registration/discovery before starting a flow.
-	client, err := findClient(ctx, clients, normalized)
+	client, err := findClient(ctx, clients, normalized, merged.ClientRef)
```
[the remark this answers is not in the document]
Comment (dev-arya23) [reply oauth/authorization.go:111]: **Independent validation of this automated finding:**

I traced the code paths and **the observation is factually correct** — `authorizationURL` does not populate `ps.ClientRef` when constructing `PendingAuthState`, and `handleCallback` builds `TokenKey` without `ps.ClientRef`. If a static client (`ClientRef != ""`) initiates an authorization flow today, the token would land in the `""` slot.

However, I believe this is **out of scope for this PR** and does not warrant blocking it:

1. **This PR's scope** (AUTH-0010, Commit 2) is explicitly "thread `ClientRef` through **registration and lookup**" — the authorization+callback flow threading is not claimed in the description or the referenced proposal commit.
2. **The `PendingAuthState.ClientRef` field already exists** (added in PR #33 / AUTH-0009), and `TokenKey.ClientRef` is already present — the wiring gap is localized to `authorizationURL` and `handleCallback`, which are untouched by this PR except for updating the `findClient` call.
3. **Authorization flow for static clients is not yet a production path** — `RegisterStaticClient` just landed in this PR. The authorization threading would naturally be a follow-up commit.

**Severity assessment**: This is a **real gap** in the multi-client support story, but it's a **follow-up task**, not a regression introduced by this PR. The pre-existing authorization code was already `ClientRef`-unaware before this change.

@Prabhjot-Sethi — is the authorization+callback `ClientRef` threading planned as a later commit in the series? If so, this is expected. If AUTH-0010 was meant to cover it, then it's a miss.
